packages/contracts/contracts/L1/ValidatorPool.sol

```diff
@@ -105,72 +93,196 @@ contract ValidatorPool is ReentrancyGuardUpgradeable, Semver {
+        require(msg.sender == address(L2_ORACLE), "ValidatorPool: sender is not L2OutputOracle");
+        require(_amount > 0, "ValidatorPool: amount cannot be 0");
+        require(_expiresAt > block.timestamp, "ValidatorPool: expiration time should be future");
+
+        Types.Bonding storage bonding = bonds[_outputIndex];
+        require(bonding.expiresAt == 0, "ValidatorPool: exists bonding of given output index");
+
+        // Unbond someone's bonding if available.
+        tryUnbond();
+
+        address submitter = getOutputSubmitter(_outputIndex);
+        _decreaseBalance(submitter, _amount);
+
+        bonding.amount = _amount;
+        bonding.expiresAt = _expiresAt;
+
+        emit Bonded(submitter, _outputIndex, _amount, _expiresAt);
+    }
+
+    /**
+     * @notice Increases the amount in a bonding corresponding to the given output index.
+     *         This function is called when creating challenge.
+     *
+     * @param _outputIndex Index of L2 checkpoint output.
+     */
+    function increaseBond(address _challenger, uint256 _outputIndex) external payable {
+        Types.Bonding storage bonding = bonds[_outputIndex];
+        require(bonding.expiresAt > 0, "ValidatorPool: bonding is not exists");
+
+        _decreaseBalance(_challenger, bonding.amount);
+        bonding.amount = bonding.amount * 2;
+
+        emit Increased(_challenger, _outputIndex, bonding.amount);
+    }
 
-        if (!validators[msg.sender]) {
-            validators[msg.sender] = true;
-            emit NewValidator(msg.sender);
+    /**
+     * @notice Attempt to unbond. Reverts if unbonding is not possible.
+     */
+    function unbond() external nonReentrant {
+        bool released = tryUnbond();
+        require(released, "ValidatorPool: no bonding that can be unbond");
+    }
+
+    /**
+     * @notice Attempts to unbond a bonding corresponding to nextUnbondOutputIndex
+               and returns whether the unbonding was successful.
+     *
+     * @return Whether a bonding has been successfully unbonded.
+     */
+    function tryUnbond() internal returns (bool) {
+        uint256 outputIndex = nextUnbondOutputIndex;
+        address recipient = getOutputSubmitter(outputIndex);
+
+        Types.Bonding storage bonding = bonds[outputIndex];
+        if (block.timestamp >= bonding.expiresAt && bonding.amount > 0) {
+            _increaseBalance(recipient, bonding.amount);
+            nextUnbondOutputIndex = nextUnbondOutputIndex + 1;
+            emit Unbonded(outputIndex, recipient, bonding.amount);
+            bonding.amount = 0;
+
+            return true;
         }
 
-        emit Deposit(msg.sender, msg.value, newDeposits);
+        return false;
     }
 
-    function withdraw(uint256 amount) external nonReentrant {
-        uint256 deposits = depositsOf[msg.sender];
-        require(deposits >= amount, "ValidatorPool: insufficient deposits");
+    function _increaseBalance(address _validator, uint256 amount) private {
```

Comment:
I don't think it's necessary to separate the process of adding or deleting a validator because it only happens with balance changes. 
As you said, I think renaming it would help make it clearer. Do you have any ideas for renaming?

Reply:
Ok. It's good not to separate it. But I don't have any good idea of appropriate name of this function. Let's leave it for now.